onnxruntime/contrib_ops/cpu/quantization/quant_gemm.cc

```diff
@@ -22,15 +23,14 @@ class QGemm : protected GemmBase, public MatMulIntegerBase {
     const auto& b_shape = b ? b->Shape() : b_shape_;
 
     const auto* c = context->Input<Tensor>(IN_C);
-    GemmHelper helper(a->Shape(), trans_A_ != CblasNoTrans,
-                      b_shape, trans_B_ != CblasNoTrans,
-                      c != nullptr ? c->Shape() : TensorShape({}));
-    if (!helper.State().IsOK())
-      return helper.State();
+    std::unique_ptr<GemmHelper> helper;
+    ORT_RETURN_IF_ERROR(GemmHelper::Create(a->Shape(), trans_A_ != CblasNoTrans,
```

Comment:
The code before this change uses stack allocation, which costs nothing. This is introducing extra mallocs, and extra pointer chases just for marginal style improvement. ORT is a high performance library after all. We should not do this.

Reply:
Our code should have a consistent coding style. No other class use a "Status" to return errors in its constructors. And this function already has a lot of heap allocations. Adding one would not have any observable performance impact.

Reply:
I would look forward to talking to him when he is back. And here is where I got the suggestion from: https://abseil.io/tips/42 
I doubt anyone would think the old code looks good:
```
if (!helper.State().IsOK())
      return helper.State();
```
Nobody use such a way to check if a construction was successful or not. Even an Init() function would be better.

Reply:
If someone could give me a good example, then I follow the example. That would be the best. Because it helps me grow.

Reply:
I am looking at it.

Reply:
> Why we allocate a ptr on the heap? Just declare it on the stack.

The std::unique_ptr<Tensor> is more like std::optional<Tensor>.

Reply:
The above problem has been addressed in another PR:  #16280